### PR TITLE
kotlin-native: init at 1.5.31

### PR DESCRIPTION
**Pull request by fabianhjr**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by fabianhjr** [comment]
Supersedes #135454 with better reuse and darwin support.

File: pkgs/development/compilers/kotlin/default.nix [hunk 1]
```diff
@@ -1,43 +1,100 @@
...
+    inherit version;
 
-  installPhase = ''
-    mkdir -p $out
-    rm "bin/"*.bat
-    mv * $out
+    src = let
+      platform = if stdenv.isLinux then
+        "linux"
+      else if stdenv.isDarwin then
```
Comment (SubhrajyotiSen) [inline pkgs/development/compilers/kotlin/default.nix:56]: Won't it be fine if we assume that if the system isn't Linux, it is Darwin?
We can skip the empty `else` block that way. Similarly for other checks as well

Comment (fabianhjr) [reply pkgs/development/compilers/kotlin/default.nix:56]: I would expect someone to help with the windows packaging. UnU

It seems different so I didn't want to add it without testing.

Comment (fabianhjr) [reply pkgs/development/compilers/kotlin/default.nix:56]: Not really, the only thing I can do is add the hashes and let it run as is.

File: pkgs/development/compilers/kotlin/default.nix [hunk 2]
```diff
@@ -1,43 +1,100 @@
...
+  homepage = "https://kotlinlang.org/";
+  license = lib.licenses.asl20;
+  description = "A modern programming language that makes developers happier";
+  longDescription = ''
+    Kotlin is a modern but already mature programming language aimed to make
+    developers happier. It’s concise, safe, interoperable with Java and other
+    languages, and provides many ways to reuse code between multiple platforms
+    for productive programming.
+  '';
+in {
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/kotlin/default.nix:14]: Why are those variables in the let in block?

File: pkgs/development/compilers/kotlin/default.nix [hunk 3]
```diff
@@ -1,43 +1,100 @@
...
+  default = stdenv.mkDerivation rec {
+    pname = "kotlin";
+    inherit version;
+
+    src = fetchurl {
+      url =
+        "https://github.com/JetBrains/kotlin/releases/download/v${version}/kotlin-compiler-${version}.zip";
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/kotlin/default.nix:21]:
```suggestion
      url = "https://github.com/JetBrains/kotlin/releases/download/v${version}/kotlin-compiler-${version}.zip";
```

File: pkgs/development/compilers/kotlin/default.nix (hunk 1 above)
Comment (SuperSandro2000) [reply pkgs/development/compilers/kotlin/default.nix:56]: > Won't it be fine if we assume that if the system isn't Linux, it is Darwin?

No, we can't assume that.

File: pkgs/development/compilers/kotlin/default.nix (hunk 2 above)
Comment (fabianhjr) [reply pkgs/development/compilers/kotlin/default.nix:14]: As duplication and being readable upon opening the file.

File: pkgs/development/compilers/kotlin/default.nix (hunk 1 above)
Comment (fabianhjr) [reply pkgs/development/compilers/kotlin/default.nix:56]: Is there a switch like statement on platform type?

File: pkgs/development/compilers/kotlin/default.nix [hunk 4]
```diff
@@ -1,43 +1,100 @@
...
+
+    installPhase = ''
+      mkdir -p $out
+      rm "bin/"*.bat
+      mv * $out
 
-  src = fetchurl {
-    url = "https://github.com/JetBrains/kotlin/releases/download/v${version}/kotlin-compiler-${version}.zip";
-    sha256 = "sha256-Je69ubsuFl5LqO+/j/lDxF1Pw52//CwcqgWejdgTZ18=";
+      for p in $(ls $out/bin/) ; do
```
[the remark this answers is not in the document]
Comment (fabianhjr) [reply pkgs/development/compilers/kotlin/default.nix:33]: Thanks, will do that change though I need to rebase first.

File: pkgs/top-level/all-packages.nix [hunk 5]
```diff
@@ -12006,7 +12006,9 @@ with pkgs;
 
   koka = haskell.lib.justStaticExecutables (haskellPackages.callPackage ../development/compilers/koka { });
 
-  kotlin = callPackage ../development/compilers/kotlin { };
+  kotlin = (callPackage ../development/compilers/kotlin { }).default;
+
+  kotlin-native = (callPackage ../development/compilers/kotlin { }).native;
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:12011]:
```suggestion
  inherit (callPackage ../development/compilers/kotlin { }) kotlin kotlin-native;
```

File: pkgs/development/compilers/kotlin/default.nix [hunk 6]
```diff
@@ -1,43 +1,100 @@
...
+      inherit homepage license description;
+      longDescription = ''
+        Kotlin/Native is a technology for compiling Kotlin code to native
+        binaries, which can run without a virtual machine. It is an LLVM based
+        backend for the Kotlin compiler and native implementation of the Kotlin
+        standard library.
+      '';
+      maintainers = with lib.maintainers; [ fabianhjr malvo ];
+      # Missing non-x86_64-linux (specially armv7a/aarch64)
+      platforms = with lib.platforms; [ "x86_64-linux" ] ++ darwin;
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/kotlin/default.nix:97]:
```suggestion
      platforms = [ "x86_64-linux" ] ++ lib.platforms.darwin;
```

File: pkgs/development/compilers/kotlin/default.nix [hunk 7]
```diff
@@ -1,43 +1,100 @@
...
+
+    nativeBuildInputs = [ makeWrapper ];
+
+    installPhase = ''
+      mkdir -p $out
+      mv * $out
+
+      wrapProgram $out/bin/run_konan --prefix PATH ":" ${
+        lib.makeBinPath [ jre ]
+      }
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/kotlin/default.nix:85]:
```suggestion
      wrapProgram $out/bin/run_konan --prefix PATH ":" ${lib.makeBinPath [ jre ]}
```

File: pkgs/development/compilers/kotlin/default.nix (hunk 1 above)
Comment (SuperSandro2000) [reply pkgs/development/compilers/kotlin/default.nix:55]: more like https://github.com/NixOS/nixpkgs/blob/956b9023e1e0d5d20eceda0e9493311ef57361d0/pkgs/applications/networking/instant-messengers/vk-messenger/default.nix#L22

File: pkgs/development/compilers/kotlin/default.nix [hunk 8]
```diff
@@ -1,43 +1,100 @@
...
+    propagatedBuildInputs = [ jre ];
+    nativeBuildInputs = [ makeWrapper unzip ];
+
+    installPhase = ''
+      mkdir -p $out
+      rm "bin/"*.bat
+      mv * $out
+
+      for p in $(ls $out/bin/) ; do
+        wrapProgram $out/bin/$p --prefix PATH ":" ${jre}/bin ;
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/kotlin/default.nix:27]:
```suggestion
        wrapProgram $out/bin/$p --prefix PATH ":" ${lib.makeBinPath [ jre ]}
```

File: pkgs/development/compilers/kotlin/default.nix [hunk 9]
```diff
@@ -1,43 +1,100 @@
...
+      url = "https://github.com/JetBrains/kotlin/releases/download/v${version}/kotlin-compiler-${version}.zip";
+      sha256 = "sha256-ZhERKG8+WsBqrzqUA9hp2alqF2tisUGBS+YmpHJJ/p4=";
+    };
+
+    propagatedBuildInputs = [ jre ];
+    nativeBuildInputs = [ makeWrapper unzip ];
+
+    installPhase = ''
+      mkdir -p $out
+      rm "bin/"*.bat
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/kotlin/default.nix:23]:
```suggestion
      rm bin/*.bat
```

**Comment by fabianhjr** [comment]
Thanks for the review @SuperSandro2000, amending the commit now.

File: pkgs/development/compilers/kotlin/default.nix (hunk 9 above)
Comment (fabianhjr) [reply pkgs/development/compilers/kotlin/default.nix:23]: The suggested change of:

```
-      rm "bin/"*.bat
+      rm bin/*.bat
```

Is on unchanged code from the current HEAD of kotlin; I don't think it is appropriate to include extra unrelated to adding kotlin-native changes.

**Comment by fabianhjr** [comment]
Built locally ran -help / -version commands.

All changes minus one applied, thanks @SuperSandro2000  for the pointer of how vk-messenger make a switch-like control structure for multiple platforms, the nix-expression in way better now.

**Review by SubhrajyotiSen — APPROVED** [review]

**Comment by AndersonTorres** [comment]
Now I have read the full file, it is very confusing.
It would be way clearer if each derivation had its own file.

**Comment by fabianhjr** [comment]
The only thing that pushed me to a single file is that Kotlin and Kotlin-native are release together from the same repo so from my point of view their versions should be kept in sync.

**Comment by AndersonTorres** [comment]
It is a bit more subtle.

First, Nixpkgs do not follow the Guix way of "putting everything in a same file".

Bundling a bunch of related things in a same _file_ is not something very readable on Nix language (and arguably not in the Guix DSL either).

It is preferrable to put them in the same _directory_ instead. Something like this:

```nix
kotlin = callPackage ../development/compilers/default.nix {}; /* default is optional */
kotlin-native = callPackage ../development/compilers/native.nix {};
``` 

Another argument for the separation of them in the two files is: `kotlin-native` is precompiled whereas `kotlin` is not. 

Therefore, they are handled a bit differently.
Sometimes precompiled programs need to be patched and post-processed in order to find libraries, configs etc.
On the other hand, locally compiled programs need a full compilation suite.

Also, while it is true that the Kotlin team releases both the source and the precompiled at the same time, it is not always the case in all scenarios.

Further, compiling from sources is not always free of failure. 
What if, for any reason, the source code does not compile in Nix but the precompiled does?
Binding them to a same file, even sharing common variables, can render both unusable whereas only one of them is indeed unusable.

**Comment by fabianhjr** [comment]
> Another argument for the separation of them in the two files is: kotlin-native is precompiled whereas kotlin is not.

Both are precompiled though. (Admittedly it would be better to build rather than fetching the built artifacts)

**Comment by AndersonTorres** [comment]
None of them are generated from human-readable source code? I didn't know.

However I think the argument still holds water.

**Pull request closed without merging**